File: day 4/AOC4.py
## part 1
def count_word_occurrences(filename, word="XMAS"):
    # Read the grid from the file
    with open(filename, 'r') as file:
        grid = [line.strip() for line in file if line.strip()]

    rows, cols = len(grid), len(grid[0])
    word_length = len(word)
    directions = [
        (0, 1),   # Right
        (1, 0),   # Down
        (1, 1),   # Diagonal Down-Right
        (1, -1),  # Diagonal Down-Left
        (0, -1),  # Left
        (-1, 0),  # Up
        (-1, -1), # Diagonal Up-Left
        (-1, 1),  # Diagonal Up-Right
    ]

    def is_valid(r, c):
        return 0 <= r < rows and 0 <= c < cols

    def count_from(r, c, dr, dc):
        """Count occurrences of the word starting from (r, c) in direction (dr, dc)."""
        for i in range(word_length):
            nr, nc = r + dr * i, c + dc * i
            if not is_valid(nr, nc) or grid[nr][nc] != word[i]:
                return 0
        return 1

    total_count = 0

    # Iterate over the grid and check for the word in all directions
    for r in range(rows):
        for c in range(cols):
            for dr, dc in directions:
                total_count += count_from(r, c, dr, dc)

    return total_count
## part 2
def count_x_mas_occurrences(filename):
    # Read the grid from the file
    with open(filename, 'r') as file:
        grid = [line.strip() for line in file if line.strip()]

    rows, cols = len(grid), len(grid[0])
    patterns = ["MAS", "SAM"]  # Possible "MAS" patterns, including reversed
    x_mas_count = 0

    def is_valid(r, c):
        return 0 <= r < rows and 0 <= c < cols

    def check_x_mas(center_r, center_c):
        """Check for an X-MAS centered at (center_r, center_c)."""
        # Coordinates for the diagonals
        diag1 = [(center_r - 1, center_c - 1), (center_r, center_c), (center_r + 1, center_c + 1)]
        diag2 = [(center_r - 1, center_c + 1), (center_r, center_c), (center_r + 1, center_c - 1)]

        # Ensure all positions are valid
        if not all(is_valid(r, c) for r, c in diag1 + diag2):
            return 0

        # Extract characters for both diagonals
        chars_diag1 = ''.join(grid[r][c] for r, c in diag1)
        chars_diag2 = ''.join(grid[r][c] for r, c in diag2)

        # Check for valid X-MAS patterns
        for pattern1 in patterns:
            for pattern2 in patterns:
                if chars_diag1 == pattern1 and chars_diag2 == pattern2:
                    return 1
        return 0

    # Iterate over the grid to find X-MAS patterns
    for r in range(1, rows - 1):  # Avoid edges
        for c in range(1, cols - 1):  # Avoid edges
            x_mas_count += check_x_mas(r, c)

    return x_mas_count
# Example usage:
filename = "xmas.txt"  # Replace with the actual filename
# result = count_word_occurrences(filename)
# print(f"The word XMAS appears {result} times in the word search.")
result = count_x_mas_occurrences(filename)
print(f"The X-MAS pattern appears {result} times in the word search.")

File: 2024/day 4/AOC4.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func countXMasOccurrences(filename string) int {
	file, err := os.Open(filename)
	if err != nil {
		fmt.Printf("Error opening file: %v\n", err)
		return 0
	}
	defer file.Close()

	var grid []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if line != "" {
			grid = append(grid, line)
		}
	}

	rows := len(grid)
	if rows == 0 {
		return 0
	}
	cols := len(grid[0])
	patterns := []string{"MAS", "SAM"}
	xMasCount := 0

	isValid := func(r, c int) bool {
		return r >= 0 && r < rows && c >= 0 && c < cols
	}

	checkXMas := func(centerR, centerC int) int {
		// Coordinates for the diagonals
		diag1 := [][2]int{{centerR - 1, centerC - 1}, {centerR, centerC}, {centerR + 1, centerC + 1}}
		diag2 := [][2]int{{centerR - 1, centerC + 1}, {centerR, centerC}, {centerR + 1, centerC - 1}}

		// Check if all positions are valid
		for _, pos := range append(diag1, diag2...) {
			if !isValid(pos[0], pos[1]) {
				return 0
			}
		}

		// Extract characters for both diagonals
		charsDiag1 := ""
		for _, pos := range diag1 {
			charsDiag1 += string(grid[pos[0]][pos[1]])
		}
		charsDiag2 := ""
		for _, pos := range diag2 {
			charsDiag2 += string(grid[pos[0]][pos[1]])
		}

		// Check for valid X-MAS patterns
		for _, pattern1 := range patterns {
			for _, pattern2 := range patterns {
				if charsDiag1 == pattern1 && charsDiag2 == pattern2 {
					return 1
				}
			}
		}
		return 0
	}

	// Iterate over the grid to find X-MAS patterns (avoid edges)
	for r := 1; r < rows-1; r++ {
		for c := 1; c < cols-1; c++ {
			xMasCount += checkXMas(r, c)
		}
	}

	return xMasCount
}

func main() {
	filename := "xmas.txt"
	result := countXMasOccurrences(filename)
	fmt.Printf("The X-MAS pattern appears %d times in the word search.\n", result)
}
